hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheProxy.java

```diff
@@ -216,9 +219,49 @@ public V getAndPut(K key, V value, ExpiryPolicy expiryPolicy) {
     public void putAll(Map<? extends K, ? extends V> map, ExpiryPolicy expiryPolicy) {
         ensureOpen();
         validateNotNull(map);
-        // TODO implement batch putAll
-        for (Map.Entry<? extends K, ? extends V> entry : map.entrySet()) {
-            put(entry.getKey(), entry.getValue(), expiryPolicy);
+
+        int partitionCount = partitionService.getPartitionCount();
+
+        try {
+            List<Future> futures = new ArrayList<Future>(partitionCount);
+            List<Map.Entry<Data, Data>>[] entriesPerPartition = new List[partitionCount];
+
```

Comment:
Hmm, we are allocating an array that's 271 elements long even if there's data from one partition only.